src/qps/interface/qps.c

```diff
@@ -1160,75 +1163,46 @@ PetscErrorCode QPSMonitor(QPS qps, PetscInt it, PetscReal rnorm)
+-  monitordestroy - [optional] routine that frees monitor context (may be NULL)
 
    Calling Sequence of monitor:
-$     monitor (QPS qps, int it, PetscReal rnorm, void *mctx)
+$     monitor (QPS qps, PetscInt it, PetscReal rnorm, void *mctx)
 
 +  qps - iterative context obtained from QPSCreate()
 .  it - iteration number
-.  rnorm - (estimated) 2-norm of (preconditioned) residual
+.  rnorm - estimated 2-norm of the residual
 -  mctx  - optional monitoring context, as set by QPSMonitorSet()
 
    Options Database Keys:
 +  -qps_monitor - sets QPSMonitorDefault()
-.  -qps_monitor_true_residual    - sets QPSMonitorTrueResidualNorm()
-.  -qps_monitor_max    - sets QPSMonitorTrueResidualMaxNorm()
-.  -qps_monitor_lg_residualnorm    - sets line graph monitor,
-                             uses QPSMonitorLGResidualNormCreate()
-.  -qps_monitor_lg_true_residualnorm   - sets line graph monitor,
-                             uses KSPMonitorLGResidualNormCreate()
-.  -qps_monitor_singular_value    - sets QPSMonitorSingularValue()
--  -qps_monitor_cancel - cancels all monitors that have
-         been hardwired into a code by
-         calls to QPSMonitorSet(), but
-         does not cancel those set via
-         the options database.
+-  -qps_monitor_cancel - Cancels all monitors set from options or with QPSMonitorSet().
 
    Notes:
-   The default is to do nothing.  To print the residual, or preconditioned
-   residual if QPSSetNormType(qps,QPS_NORM_PRECONDITIONED) was called, use
-   QPSMonitorDefault() as the monitoring routine, with a null monitoring
-   context.
+   The default is to do nothing. To print the residual, use
+   QPSMonitorDefault() as the monitoring routine, with a null monitoring context.
    
    Several different monitoring routines may be set by calling
    QPSMonitorSet() multiple times; all will be called in the
    order in which they were set.
    
-   
    Level: beginner
 
-.keywords: QPS, set, monitor
-
-.seealso: QPSMonitorDefault(), QPSMonitorLGResidualNormCreate(), QPSMonitorCancel()
+.seealso: QPSMonitorDefault(), QPSMonitorCancel()
 @*/
 PetscErrorCode  QPSMonitorSet(QPS qps,PetscErrorCode (*monitor)(QPS,PetscInt,PetscReal,void*),void *mctx,PetscErrorCode (*monitordestroy)(void**))
 {
-//   PetscInt       i;
-
    PetscFunctionBegin;  
+   if (!monitor) SETERRQ(PetscObjectComm((PetscObject)qps),PETSC_ERR_ARG_NULL,"Monitor function must be specified");
+
    /* verify the number of monitors */
    if (qps->numbermonitors >= MAXQPSMONITORS) SETERRQ(PetscObjectComm((PetscObject)qps),PETSC_ERR_ARG_OUTOFRANGE,"Too many QPS monitors set");
 
-   /*
-   for (i=0; i<qps->numbermonitors;i++) {
-     if (monitor == qps->monitor[i] && monitordestroy == qps->monitordestroy[i] && mctx == qps->monitorcontext[i]) {
-       if (monitordestroy) {
-         (*monitordestroy)(&mctx);
-       }
-       return(0);
-     }
-   }
-   */
```

Comment:
Not sure why this was commented out, but (dropping the second if) this would ensure that two exactly same monitors are not set. I think that is reasonable.

Reply:
Comparing the destroy function and context looks very weird to me. They don't define the monitor, and it's unlikely exactly the same context pointer would be passed. Also, if the user sets the same monitor twice, it is his wish :-) What about just making sure that `-qps_monitor` sets the monitor just once?

Reply:
I looked into it and if you check SetMonitor routines (e.g., [TSMonitorSet()](https://www.mcs.anl.gov/petsc/petsc-current/src/ts/interface/ts.c.html#TSMonitorSet)) a similar check is done using [PetscMonitorCompare()](https://www.mcs.anl.gov/petsc/petsc-current/src/sys/classes/viewer/interface/viewers.c.html#PetscMonitorCompare).
The docs say that `Checks if two monitors are identical; if they are then it destroys the new one`, but they are identical iff ctx == newctx, so that destroy on newctx also destroys ctx or am I missing something?

I think it does not make sense to call the same monitors (having the same ctx) more than once. On the other hand, it makes sense to set multiple identical monitor routines with different contexts.

Reply:
> The docs say that Checks if two monitors are identical; if they are then it destroys the new one, but they are identical iff ctx == newctx, so that destroy on newctx also destroys ctx or am I missing something?

Yeah, I think it doesn't make any sense. Could be filed as a bug actually...

I don't think we should necessarily duplicate the PETSc behavior which is IMHO odd.

> I think it does not make sense to call the same monitors (having the same ctx) more than once. On the other hand, it makes sense to set multiple identical monitor routines with different contexts.

What is the difference between two monitors with the same monitor function and different context? It's still the same monitor and should print exactly the same. The context is just a helper thing.

---
Aside, I think PETSc overuses this model (callback + context + destroy). Sometimes it would be more elegant and less error-prone to have a PetscObject which has its own methods. An extreme case are some functions related to discretization, see https://gitlab.com/petsc/petsc/-/merge_requests/2793#note_343818944.

Reply:
> > I think it does not make sense to call the same monitors (having the same ctx) more than once. On the other hand, it makes sense to set multiple identical monitor routines with different contexts.
> 
> What is the difference between two monitors with the same monitor function and different context? It's still the same monitor and should print exactly the same. The context is just a helper thing.
> 
You could have a variable in ctx by which the monitor decides what and how it is printed. E.g., you could have a monitor on a subset of a problem (e.g., using a specific submatrix) and then it also makes sense to monitor multiple subsets with the same monitor (but different ctx).

I think this is exactly the reason why there are contexts for monitors.

Reply:
OK

Reply:
OK, actually we could just use PetscMonitorCompare(), and create a petsc merge request fixing that. I think if the contexts are identical, it should just return true but not call destroy.

I'm also not persuaded it makes sense to compare the destroy functions.

Reply:
I think we can leave the `for` loop without the `if (monitordestroy)` for now. Then this can be merged.

Reply:
Keep it commented out or uncomment?

Reply:
Uncomment, without the wrong if.

Reply:
OK